### PR TITLE
Mappable values

### LINKED ISSUE #1897: mappable query params and path params
E.g.

```
req.path.param("foo").asInt().ifPresent()
req.queryParams().first("foo").asInt().ifPresent()
```

Related to #1789 and #1867

### PR DESCRIPTION
**Pull request by tomas-langer**
Resolves #1897 (mappable query params and path params)
Related to #1789 (align mappers)

Added `Value` and `ValueProvider` to `mapper` module.
It is now used in various HTTP classes (Headers, query parameters etc.) and in Config (`ConfigValue`).

Added support for a shared `MapperManager` that can be custom registered by user.
Mapping to basic types is built-in (String -> a lot of types), including date types. Format can be overridden through context if desired (see `MapperManager.FORMAT_CLASSIFIER`).

The changes are fully backward compatible - we have to decide how to approach this (e.g. do we want to keep all the methods, or do we want to change signatures in the future?).

Created as a draft until we agree on approach.

### CODE REVIEW

**Comment by romain-grecourt** [comment]
It would be easier to have a markdown document going over the proposal.
We should describe all the various use-cases with snippets:
 - config
 - dbclient
 - media support
 - webserver query params

Highlight the complex type usages with GenericType, document the special layering of media-support.

We should also have an example that demonstrates the value of this API when combining dbclient and webserver/webclient. E.g. Show how to support a custom entity all the way from a request payload to the database.

File: common/mapper/src/main/java/io/helidon/common/mapper/BackedValue.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+ * limitations under the License.
+ */
+
+package io.helidon.common.mapper;
+
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Function;
+
+class BackedValue<T> extends EmptyValue<T> {
```
Comment (danielkec) [inline common/mapper/src/main/java/io/helidon/common/mapper/BackedValue.java:23]: Why extend the EmptyValue? Only common func seems to be name accessor

File: common/mapper/src/main/java/io/helidon/common/mapper/MapperManagerImpl.java [hunk 2]
```diff
@@ -81,6 +123,10 @@ private RuntimeException createMapperException(Object source,
                     .orElseGet(() -> {
                         GenericType<SOURCE> sourceGenericType = GenericType.create(sourceType);
                         GenericType<TARGET> targetGenericType = GenericType.create(targetType);
+                        if (targetType.isAssignableFrom(sourceType)) {
+                            // the same type
+                            return it -> it;
```
Comment (danielkec) [inline common/mapper/src/main/java/io/helidon/common/mapper/MapperManagerImpl.java:128]: Would be nice to have something like `Fuction.identity()` maybe `Mapper.noop()`?

File: common/mapper/src/main/java/io/helidon/common/mapper/SharedManagerHandler.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.helidon.common.mapper;
+
+import java.util.concurrent.atomic.AtomicReference;
+
+final class SharedManagerHandler {
+    private static final AtomicReference<MapperManager> SHARED = new AtomicReference<>(MapperManager.builder()
```
Comment (danielkec) [inline common/mapper/src/main/java/io/helidon/common/mapper/SharedManagerHandler.java:22]: Could be just volatile

File: common/mapper/src/main/java/io/helidon/common/mapper/Value.java [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+     * @param mapper a mapping function to apply to the value, if present
+     *               the mapping function
+     * @return the result of applying an {@code Optional}-bearing mapping
+     * function to the value of this {@code Optional}, if a value is present,
+     * otherwise an empty {@code Optional}
+     * @throws NullPointerException if the mapping function is null or returns
+     *                              a null result
+     * @see Optional#flatMap(Function)
+     */
+    default <U> Optional<U> flatMap(Function<? super T, Optional<U>> mapper) {
```
Comment (danielkec) [inline common/mapper/src/main/java/io/helidon/common/mapper/Value.java:232]: Why not `Function<? super T, Optional<? extends U>> mapper` ?

File: common/mapper/src/main/java/io/helidon/common/mapper/BackedValue.java (hunk 1 above)
Comment (tomas-langer) [reply common/mapper/src/main/java/io/helidon/common/mapper/BackedValue.java:23]: yes, that is the only common thing. On the other hand, why implement it twice? Something with value is a logical extension of something without a value...

File: common/mapper/src/main/java/io/helidon/common/mapper/Value.java (hunk 4 above)
Comment (tomas-langer) [reply common/mapper/src/main/java/io/helidon/common/mapper/Value.java:232]: fixed

File: common/mapper/src/main/java/io/helidon/common/mapper/SharedManagerHandler.java (hunk 3 above)
Comment (tomas-langer) [reply common/mapper/src/main/java/io/helidon/common/mapper/SharedManagerHandler.java:22]: Fixed

File: common/mapper/src/main/java/io/helidon/common/mapper/MapperManagerImpl.java (hunk 2 above)
Comment (tomas-langer) [reply common/mapper/src/main/java/io/helidon/common/mapper/MapperManagerImpl.java:128]: Fixed

**Comment by tomas-langer** [comment]
I am adding a `docs-internal/generic-mapping-2.md`

**Pull request closed without merging**